### PR TITLE
Schema/improved regex handling

### LINKED ISSUE #761: Pattern and PatternProperties keyword performance and security improvements
### Nuget Package

JsonSchema.Net

### Package Version

7.1.4

### Can this library make doing something simpler?

The PatternKeyword and the PatternPropertiesKeyword implementations both create new compiled `Regex`  instances.

According to best practices as described here: https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#static-regular-expressions regex instances that are often used should be cached for performance benefits.

To achieve this, the keywords could use the static `Regex.IsMatch` method to use the internal cache (optionally add an entry in the `readme`  file to increase the [Regex.CacheSize](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.cachesize?view=net-8.0) property.

Another options is to implement a regex cache within the JsonSchema.net library.

Also, `Regex` instances [should set a timeout](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#use-time-out-values) especially because json schema's will validate untrusted user input.

I'd be happy to create a PR for this, but that would require breaking changes to the `PatternKeyword` and the `PatternPropertiesKeyword` I think. 

### Describe alternatives you've considered.

_No response_

### Is there any other information you'd like to share regarding this enhancement?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by gregsdennis**
### Description

Updating details from #773 

### Links

Resolves #761

### Checks

- [x] I have read and understand the [Code of Conduct](https://github.com/json-everything/json-everything/blob/master/CODE_OF_CONDUCT.md) and [Contribution Guidelines](https://github.com/json-everything/json-everything/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: src/JsonSchema/PatternPropertiesKeyword.cs [hunk 1]
```diff
@@ -25,43 +25,65 @@ namespace Json.Schema;
...
+	public IReadOnlyDictionary<string, JsonSchema> PatternValues => _patternValues ??= _patternsLookup.ToDictionary(x => x.Key, x => x.Value.Schema);
+	
 	/// <summary>
-	/// If any pattern is invalid or unsupported by <see cref="Regex"/>, it will appear here.
+	/// The regex patterns of this PatternPropertiesKeyword
 	/// </summary>
-	/// <remarks>
-	/// All validations will fail if this is populated.
-	/// </remarks>
-	public IReadOnlyList<string>? InvalidPatterns { get; }
```
Comment (gregsdennis) [inline src/JsonSchema/PatternPropertiesKeyword.cs:46]: @rikbosch how is this being handled now?

File: src/JsonSchema/PatternKeyword.cs [hunk 2]
```diff
@@ -23,39 +24,44 @@ namespace Json.Schema;
...
-	/// </summary>
-	public Regex Value { get; }
-	/// <summary>
-	/// If the pattern is invalid or unsupported by <see cref="Regex"/>, it will appear here.
+	/// Returns the Regex Value of the keyword
 	/// </summary>
-	/// <remarks>
-	/// All validations will fail if this is populated.
-	/// </remarks>
-	public string? InvalidPattern { get; }
```
Comment (gregsdennis) [inline src/JsonSchema/PatternKeyword.cs:41]: Same here.

Comment (gregsdennis) [reply src/JsonSchema/PatternKeyword.cs:41]: I've added a couple tests to `master` and rebased to show this.

Comment (gregsdennis) [reply src/JsonSchema/PatternKeyword.cs:41]: If `RegexOrPattern` were to check the pattern in the constructor, would that insert the pattern into the Regex cache (what this whole change was trying to do anyway)?

> I would say that it's an illegal json schema when the regex pattern is invalid, so an exception when deserializing json to a JsonSchema sounds valid.

I want to say there were some optional tests that had valid regexes that .Net just didn't support.  It seems the only one that fits that now are the UTF-16 tests, but even those just don't evaluate correctly.

Moreover, the `.Pattern()` builder extension _does_ throw an exception on a bad pattern as well.

I think I want to change the behavior long term, but I'd still like to leave the current support for now.  I'll work something out.

Comment (gregsdennis) [reply src/JsonSchema/PatternKeyword.cs:41]: I suppose I can include this change and say in the release notes that it's fixing a bug.  Doesn't really explain why the `InvalidPattern` property was there in the first place, though.

Comment (rikbosch) [reply src/JsonSchema/PatternKeyword.cs:41]: Yeah, performing an `IsMatch` in the constructor works as well, and is actually closer to the original behavior.

**Pull request merged**